Fullscreen mode with hidden menu and wheel navigation

tobias-klein opened the pull request:
![image](https://user-images.githubusercontent.com/17739895/135656922-ec8b0e10-9660-4fab-b069-a684d62ad44c.png)

Open points:

- [x] Icons
- [x] Turning the change tag menu into a dialog during this working mode.
- [x] Add the filter and search options to the change tag menu.
- [x] Add the new tag button to the change tag menu.
- [x] Enforce popup mode for opening tagged verse lists.
- [x] Find a solution for the wheel navigation on the first verse where there is not enough space to show the full menu.
- [x] Add a loading indicator to the tag list popup dialog.
- [x] Consider _notes_ in the wheel navigation. (see #253)
- [x] Disable button in tag list popup _Open verse list in new tab_.
- [ ] Consider moving the fullscreen button into the corner of the verse list.
- [ ] Try to position the tag list popup below or above the current verse if possible.
- [x] Consider different colors in night mode
- [x] Style verse number as hyperlink
- [x] Add copy to clipboard button

tobias-klein commented:
@augustin-colesnic 
This is the current work in progress ... feel free to check it out!

augustin-colesnic commented:
late this evening i can review it. until then i see you still left the book tab above. i would love the idea of real full screen mode where the space taken by the book tab is also freed up and used by the bible text. 👌

tobias-klein commented:
@augustin-colesnic There are still quite some open points before this can be merged, see the list in the description.

Regarding the tab navigation. Even when focussing on the Bible text, I find myself often working with multiple passages in parallel. Therefore I think the possibility to switch tabs even in fullscreen mode remains important. However, we could make this configurable!

tobias-klein commented:
@augustin-colesnic Latest status visible in the screenshot in the summary. I changed it so that the tab bar is also hidden in fullscreen mode now.

tobias-klein commented:
@zhuiks Any early feedback on this? I find this new fullscreen mode particularly useful on tablets. Should we make this an option or default behavior?

zhuiks commented:
Sorry for coming late on this. Here are my thoughts:
1. Removing of the whole menu in full-screen mode makes a big sense. I definitely support it!
2. I think we don't need any switchable options for this. This should be the default behavior. Extra option will give extra confusion.
3. To "hint" a user for available touch/click options, verse number can be styled as "hyperlink":
![Знімок екрану з 2021-10-04 14-12-55](https://user-images.githubusercontent.com/3584119/135919561-ebaadde9-f55a-4ba8-b5ce-91aca18b79b8.png)

4. It's hard to fit text to the wheel navigation. I think only icons should be there. On touch/click the action should be obvious for the user with either with dialog caption or popup menu caption. Or in case of "last used tag assignment" there should be the easy way from the wheel menu to unassign the tag.
5. As an alternative to the wheel menu or option for not-fullscreen mode on verse selection/verse number click popup buttons can be displayed on top of the verse:
![Знімок екрану з 2021-10-04 13-58-05](https://user-images.githubusercontent.com/3584119/135919523-2797da8b-9e53-4f3b-b559-f246b0f72cde.png)
6. The colors of the wheel menu should be adjusted to the "night mode"
7. The wheel menu is a good place for "copy to clipboard" button #254

tobias-klein commented:
This is how the latest version looks now.

![image](https://user-images.githubusercontent.com/17739895/139645935-9c3ec731-53f1-4d9c-90b7-9510e22eafe2.png)

* Added function for editing notes
* Added function for copying verse to clipboard
* Changed icon for "Change tags" function (collection of multiple tags seems to be easier to understand)

@zhuiks @augustin-colesnic 
Feedback?

tobias-klein commented:
@zhuiks @augustin-colesnic 
This is ready for review now and I am done from a functional perspective. I'd be happy about any feedback.

zhuiks commented:
One thing that I noticed:
![Screenshot from 2021-11-08 16-28-02](https://user-images.githubusercontent.com/3584119/140828589-fdd8586d-d9c9-48ba-84f5-d40e94216a8c.png)

I think, that verse reference should be straight in the center of the circle.
I propose to align verse references on center in `main.css#L536`:
```
text-align: center;
```
`float: left` can be removed from there as it is legacy code. The parent container is a flex-box now.

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -140,6 +141,50 @@ class AppController {
+    editNoteButton.bind('click', (event) => {
+      event.stopPropagation();
+
+      if (!$(this).hasClass('ui-state-disabled')) {
```
Remove dependency on jQuery:
``` javascript
if (!this.classList.contains('ui-state-disabled')) {
```

tobias-klein replied:
Ok

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -140,6 +141,50 @@ class AppController {
+    copyButton.bind('click', (event) => {
+      event.stopPropagation();
+
+      if (!$(this).hasClass('ui-state-disabled')) {
```
Same as above:
```javascript
if (!this.classList.contains('ui-state-disabled')) {
```

tobias-klein replied:
Ok

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -140,6 +141,50 @@ class AppController {
+  }
+
+  enableVerseButtons() {
+    var currentVerseListMenu = app_controller.getCurrentVerseListMenu();
+    var editNoteButton = currentVerseListMenu.find('.edit-note-button');
+    var copyButton = currentVerseListMenu.find('.copy-clipboard-button');
+    editNoteButton[0].classList.remove('ui-state-disabled');
+    copyButton[0].classList.remove('ui-state-disabled');
```
This can be refactored in vanilla javascript without jQuery:
```javascript
    var currentVerseListMenu = app_controller.getCurrentVerseListMenu()[0];
    var editNoteButton = currentVerseListMenu.querySelector('.edit-note-button');
    var copyButton = currentVerseListMenu.querySelector('.copy-clipboard-button');
    editNoteButton.classList.remove('ui-state-disabled');
    copyButton.classList.remove('ui-state-disabled');
```

tobias-klein replied:
Ok

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -140,6 +141,50 @@ class AppController {
+  }
+
+  disableVerseButtons() {
+    var currentVerseListMenu = app_controller.getCurrentVerseListMenu();
+    var editNoteButton = currentVerseListMenu.find('.edit-note-button');
+    var copyButton = currentVerseListMenu.find('.copy-clipboard-button');
+    editNoteButton[0].classList.add('ui-state-disabled');
+    copyButton[0].classList.add('ui-state-disabled');
```
Same as above. No need for extra jQuery `find`.

tobias-klein replied:
Ok

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -480,20 +524,22 @@ class AppController {
     var platform = null;
 
     if (platformHelper.isElectron()) {
-
       platform = electronPlatform;
-
     } else if (platformHelper.isAndroid()) {
-
       platform = cordovaPlatform;
-
     }
 
+    return platform;
+  }
```
I think it's a good method to move to `ezraHelper` module.

Ideally, `app_controller` can be the main container without much logic.

tobias-klein replied:
Ok. Will move it to platform helper directly.

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -480,20 +524,22 @@ class AppController {
+    return platform;
+  }
+
+  toggleFullScreen() {
```
Should all full-screen functionality go to a separate controller/component?

tobias-klein replied:
It's literally just this one function + the initialization of the short cut (app_controller) and button (startup.js). So it's not all that much ...

tobias-klein replied:
I moved that functionality to `fullscreen_controller.js` now!

zhuiks commented on app/frontend/controllers/app_controller.js:
```diff
@@ -502,50 +548,36 @@ class AppController {
     }
+  }
+
+  async copySelectedVerseTextToClipboard() {
+    var selectedVerseText = await this.verse_selection.getSelectedVerseText();
+    this.getPlatform().copyTextToClipboard(selectedVerseText);
+  }
```
Maybe this can be a part of `verse_selection` component?

tobias-klein replied:
Ok

zhuiks reviewed (changes requested):
I have some comments :) Mostly regarding dependencies on jQuery functionality.

In general, I think, it would be nice to have wheel-nav connected in one place as a web-component instead of connecting it in main `index.html` and control with `wheelnavController`. But that's a chunk of work. So I think it's ok as you implemented it for now.

tobias-klein commented:
Thanks for your feedback @zhuiks - I'll definitely consider it!

augustin-colesnic commented:
> This is how the latest version looks now.
> 
> ![image](https://user-images.githubusercontent.com/17739895/139684244-92d18480-f915-423c-8618-4d7e5616d305.png)
> 
> * Added function for editing notes
> * Added function for copying verse to clipboard
> * Changed icon for "Change tags" (collection of multiple tags seems to be easier to understand)
> * Changed icon for "Assign last used tag"
> 
> @zhuiks @augustin-colesnic Feedback?

@tobias-klein  I love it, even if I would make the wheel less transparent, even opaque. Great job !

tobias-klein commented on app/frontend/controllers/app_controller.js:
```diff
@@ -140,6 +141,50 @@ class AppController {
   }
 
+  initVerseButtons(tabIndex=undefined) {
+    var currentVerseListMenu = app_controller.getCurrentVerseListMenu(tabIndex);
```
Ok

tobias-klein commented:
> I have some comments :) Mostly regarding dependencies on jQuery functionality.
> 
> In general, I think, it would be nice to have wheel-nav connected in one place as a web-component instead of connecting it in main `index.html` and control with `wheelnavController`. But that's a chunk of work. So I think it's ok as you implemented it for now.

Thank you, @zhuiks 
I addressed your comments with some updates.
Implementing the handling in a controller rather than a web component was more straightforward in this case.
I still have to have a look at the CSS formatting you mentioned.

@augustin-colesnic
I think I would prefer to keep the formatting as is. Making the wheel menu more opaque would make it more difficult to recognize it I feel. @zhuiks What do you think regarding this proposal?

tobias-klein commented:
@zhuiks
The only decently looking approach to fix the alignment was to apply `align-center` (vertical align of child elements) already on `verse-box` level. Note that this influences all elements of the verse box including notes.

It looks like this:
![image](https://user-images.githubusercontent.com/17739895/141651957-cbb0b79f-9465-4a25-ac94-ec2b3e04263a.png)

This certainly changes the look of the Bible text quite a bit and the question is whether we and our users are happy with this. What do you think?

tobias-klein commented:
@zhuiks From my perspective this is done. Does anything speak against merging now?

zhuiks commented:
@tobias-klein I found one little mistake 3559087
Everything else looks good. Nice job :star_struck:

tobias-klein commented:
Thanks for your review & fix, @zhuiks !
